riscv/src/register/macros.rs

```diff
@@ -397,3 +397,681 @@ macro_rules! clear_pmp {
         }
     };
 }
+
+/// Helper macro to define a CSR type.
+///
+/// This macro creates a type represents a CSR register, without defining any bitfields.
+///
+/// It is mainly used by [read_write_csr](crate::read_write_csr),
+/// [read_only_csr](crate::read_only_csr), and [write_only_csr](crate::write_only_csr) macros.
+#[macro_export]
+macro_rules! csr {
+    ($doc:expr, $ty:ident: $csr:literal, $mask:literal) => {
+        #[repr(C)]
+        #[doc = $doc]
+        #[derive(Clone, Copy, Debug, Eq, PartialEq)]
+        pub struct $ty {
+            bits: usize,
+        }
+
+        impl $ty {
+            /// Bitmask for legal bitfields of the CSR type.
+            pub const BITMASK: usize = $mask;
+
+            /// Creates a new CSR type from raw bits value.
+            ///
+            /// Only bits in the [BITMASK](Self::BITMASK) will be set.
+            pub const fn from_bits(bits: usize) -> Self {
+                Self { bits: bits & $mask }
+            }
+
+            /// Gets the raw bits value.
+            pub const fn bits(&self) -> usize {
+                self.bits & $mask
+            }
+
+            /// Gets the bitmask for the CSR type's bitfields.
+            pub const fn bitmask(&self) -> usize {
+                Self::BITMASK
+            }
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! csr_field_enum {
+    ($field_ty_doc:expr,
+     $field_ty:ident {
+         range: [$field_start:literal : $field_end:literal],
+         default: $default_variant:ident,
+         $($variant:ident = $value:expr$(,)?)+
+     }$(,)?
+    ) => {
+         #[doc = $field_ty_doc]
+         #[repr(usize)]
+         #[derive(Clone, Copy, Debug, Eq, PartialEq)]
+         pub enum $field_ty {
+             $($variant = $value),+
+         }
+
+         impl $field_ty {
+             /// Creates a new field variant.
+             pub const fn new() -> Self {
+                 Self::$default_variant
+             }
+
+             /// Attempts to convert a [`usize`] into a valid variant.
+             pub const fn from_usize(val: usize) -> Option<Self> {
+                 match val {
+                     $($value => Some(Self::$variant),)+
+                     _ => None,
+                 }
+             }
+
+             /// Converts the variant into a [`usize`].
+             pub const fn into_usize(self) -> usize {
+                 self as usize
+             }
```

Comment:
> Regarding from and into: if we can get help from From and TryFrom traits, I'd go for them instead of maintaining repetitive methods.

While I agree cutting down on repetition is good, adding these `from`/`into` methods here actually provides utility. Because [const traits](https://github.com/rust-lang/rust/issues/67792) aren't available yet, these methods allow for constructing CSR registers in `const` contexts.

We could probably workshop the name (e.g. `from_inner`), but I do think these functions are worth keeping.

Changing the inner representation type to a macro-defined type might also be a good thing to add.

If it's a deal-breaker for merging the PR, I can remove them.